### DISCUSSION TITLE
How calculate standard deviation in grid_search_forecaster() and  in backtesting_forecaster()?

### DISCUSSION THREAD

**Discussion by lucayapi** [JoaquinAmatRodrigo/skforecast#700] [Q&A] [answered]
Hello,`thanks for this great package.

In the The skforecast 0.12.1  , grid_search_forecaster() and backtesting_forecaster() doesn't return the standard deviation of metrics_value for each hyperparameter ? does it exist a way to get these values in the dataframe returned? 

```
results_grid = grid_search_forecaster(
    forecaster = forecaster,
    y = data_train["y"],
    param_grid= param_grid,
    lags_grid= lags_grid,
    steps = steps,
    refit = False,
    metric = "mean_squared_error",
    initial_train_size=int(len(data_train)*0.5),
    fixed_train_size = False,
    return_best = True,
    n_jobs = "auto",
    verbose = False
)
```

**Comment by JoaquinAmatRodrigo** [comment]
Hi @lucayapi,
Thanks for using skforecast.

The backtesting process calculates the metric using all predicted values over the entire backtesting period.  For example, let's say you have a time series with daily frequency, a forecaster that predicts 7 steps per head, and you want to backtest it over a period of 1 year. The backtesting process will generate the 365 predictions and calculate the metric with them.  This is the metric error over the whole year. Since there is only one value, no standard deviation can be calculated.

When using hyperparameter search (such as grid search), the process will select the configuration that results in a lower error over the entire period.

I hope this explanation helps.

**Comment by lucayapi** [comment]
Thank you, 
more details about my question
if you look at the backtesting below we have 8 partitions so the metric is calculated on each validation set of each partition. 
we therefore have 8 values for each metric.

- the backtest error for what I understand is the average of the 8 values of each partition. and I agree that it is also equal to error metric of the entire backtesting period.

- But we can calculate a standard deviation of the values of each partition ( 8 values here). It measures how much the metric varies on each partition.

I hope I’m clear enough.

![image](https://github.com/JoaquinAmatRodrigo/skforecast/assets/72395060/ed33335d-12e0-40f9-b1ca-7a683cc0036b)

**Comment by JoaquinAmatRodrigo** [comment]
Not exactly. The goal of backtesting is not the same as traditional cross-validation.

In backtesting, folds are created to mimic what would happen if the model were run in production (availability of historical data and predicted steps). For each fold, predictions are made and stored, but no metric is calculated. Once the backtesting process is complete, the predictions of each fold are concatenated (since the folds are temporally ordered) and the metric is calculated for the entire period.

If the backtesting period is the year 2024, the result would be: The metric of the model for the year 2024 is xxx.

In your example, if you change the number of steps from 10 to 5, the backtesting process will generate twice as many folds, but the backtesting period will still have the same total length.

**Reply by lucayapi** [reply]
Okay I understand better thank you for your explanation !

**Reply by lucayapi** [reply]
Looking at what happens in each partition seems relevant in production though.
We would like to know if the performance of the model remains stable in time after the addition of new data and for different validations set.

**Reply by JoaquinAmatRodrigo** [reply, marked as answer]
Since the backtesting process returns the predictions for the entire backtesting period, I would recommend that you calculate the desired metric for whatever partition you find interesting (n steps, weeks, months...). For example, you might want to know "the evolution of the weekly error over a whole year".

However, this would only be possible for backtesting, not for grid search, since the latter does not return the predictions of each candidate model. If you need to compare multiple models, you would have to run backtesting on each of them independently.
